File: OOPproject/app/src/main/java/com/gorkemozcan/oopproject/MainActivity.java
package com.gorkemozcan.oopproject;

import static com.gorkemozcan.oopproject.HouseDecor.roomName;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        /*
        User myUser = new User();
        myUser.name = "Atil";
        myUser.job = "Instructor";

        User newUser = new User();
        newUser.name = "Kirk";
        newUser.job = "Musician";
        */

        User myUser = new User("Atil", "Instructor");
        System.out.println(myUser.information());
        System.out.println(myUser.name);

        //Encapsulation
        Musician james = new Musician("James ", "Guitar ", 50);
        int age = james.getAge();
        String name = james.getName();
        String instrument = james.getInstrument();
        System.out.println(age + name + instrument);
        james.setAge(70, "Gorkem");// Musicians.java sınıfında password tanımladık
        //eger set age kısmında bu password dogru girilmezse yapılacak olan değişim işleme tabi tutulmuyor.
        james.getAge();
        System.out.println(james.getAge());

        //Inheritance
        SuperMusician lars = new SuperMusician("lars", "Drums", 55);
        System.out.println(lars.sing());
        System.out.println(lars.getAge());
        SuperMusician gorkem = new SuperMusician("James", "Guitar", 50);
        System.out.println(gorkem.sing());
        System.out.println(gorkem.getInstrument());


        //Polymorphism

        //static polymorphism
        Mathematics mathematics = new Mathematics();
        System.out.println(mathematics.sum());
        System.out.println(mathematics.sum(5, 3));
        System.out.println(mathematics.sum(5, 3, 4));

        //dynamic polymorphism
        Animal myAnimal = new Animal();
        myAnimal.sing();
        Dog barley = new Dog();
        barley.test();
        barley.sing();

        //Abstract
        /*
        abstract sınıflar diğer sınıflar extend etsin diye oluşturulan sınıflardır.
         */

        Piano myPiano = new Piano();
        myPiano.brand = "Yamaha";
        myPiano.digital = true;
        myPiano.info();




    }
}

File: KotlinOOPProject/app/src/main/java/com/gorkemozcan/kotlinoopproject/MainActivity.kt
package com.gorkemozcan.kotlinoopproject

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle

class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        //Constructor

        var myUser = User("James", 50)
        myUser.name = "Lars"
        myUser.age = 60
        println(myUser.age)
        println(myUser.name)
        println(myUser.information())

        // encapsulation

        var james = Musician("james","guitar", 55)
        //james.age = 60
        println(james.age)

        println(james.returnBandName("Gorkem"))
        println(james.returnBandName("Kirk"))

        //inheritance

        var lars = SuperMusician("Lars", "Drums", 65)
        println(lars.name)
        println(lars.returnBandName("Gorkem"))
        lars.sing()

        //polymorphism
        //static polymorphism

        var mathematics = Mathematics()
        println(mathematics.sum())
        println(mathematics.sum(3,4))
        println(mathematics.sum(3,4,5))
        //dynamic polymorphism

        val animal = Animal()
        animal.sing()


        val barley = Dog()
        barley.test()
        barley.sing()


        //Abstract & interface


        //var myPeople = People() -> objesi oluşturulamaz

        var myPiano = Piano()
        myPiano.brand = "Yamaha"
        myPiano.digital = false
        println(myPiano.roomName)
        myPiano.info()


        //Lambda Expressions

        fun printString(myString : String)
        {
            println(myString)
        }

        printString("My Test String")

        val testString = {myString : String -> println(myString)}
        testString("My Lambda String")


        val multiplyLambda = {a: Int, b:Int  -> a * b}

        println(multiplyLambda(3,5))


        val multiplyLambda2 : (Int, Int) -> Int = {a,b -> a*b}
        println(multiplyLambda2(5,5))


        //asyncchrnous -> asenkronize

        //callback function, listener function, completion function

        fun downloadMusicians(url: String, completion : () -> Unit) //boş bir şey boş bir şey döndürecek ki içerde Url'den indirdiğimiz verilerin indiğinden emin olup işleme devam edebilelim
        {
            //url -> download
            //data
            completion()
        }

        downloadMusicians("metallica.com", {
            println("completed && ready")
        })



    }
}